#!/usr/bin/env Rscript


get_args <- function() {
    ### options: aggr
    option_list = list(
        optparse::make_option(c("-a", "--aggr"), action="store", default=NA, type="character", help="aggr.csv generated by make_aggregation_table()"),
        optparse::make_option(c("--tss"), action="store", default=4, type="numeric", help="TSS enrichment cutoff"),
        optparse::make_option(c("-o", "--output"), action="store", default="archr_metadata.tsv.gz", type="character", help="output file for metadata"),
        optparse::make_option(c("--genome"), action="store", default="hg38", type="character", help="Genome identifier passed to ArchR::addArchRGenome()"),
        optparse::make_option(c("-f", "--force"), dest="force", action="store_true", default=TRUE, help="Overwrite arrow files if already created."),
        optparse::make_option(c("-n", "--no-force"), dest="force", action="store_false", help="Don't overwrite arrow files if already created."),
        optparse::make_option(c("-g", "--gene-annotation"), action="store", default=NA, type="character", dest="gene_annotation", help="archr gene annotation (RDS)"),
        optparse::make_option(c("-r", "--refdata"), action="store", default=NA, type="character", help="Path to 10X style refdata for generating tile matrix"))
    return(optparse::parse_args(optparse::OptionParser(option_list=option_list)))
}

run <- function(aggr, tss, output, refdata=NA, gene_annotation=NA, help=FALSE, force=TRUE, genome="hg38", threads=as.integer(Sys.getenv("OMP_NUM_THREADS", "16"))) {
    require(ArchR)
    addArchRGenome(genome)
    addArchRThreads(threads)
    df = read.csv(aggr)
    if (is.na(gene_annotation) | !file.exists(gene_annotation)) {
        gene_annotation = getGeneAnnotation()
    } else {
        gene_annotation = readRDS(gene_annotation)
    }
    ArrowFiles = createArrowFiles(setNames(df$atac_fragments, df[[1]]), minTSS=tss, geneAnnotation=gene_annotation, force=force)
    proj = ArchRProject(ArrowFiles, "ArchR", copyArrows=FALSE, geneAnnotation=gene_annotation)
    data.table::fwrite(as.data.frame(getCellColData(proj)), output, row.names=TRUE, sep="\t")
    genome = ArchR::getArchRGenome(genomeAnnotation=TRUE)
    ###
    atac_seqnames = Reduce(intersect, lapply(df$atac_fragments, Rsamtools::seqnamesTabix))
    if (!is.na(refdata) & file.exists(paste0(refdata, "/star/chrNameLength.txt"))) {
        cs = read.table(paste0(refdata, "/star/chrNameLength.txt"), sep="\t")
        cs = with(cs[cs$V1 %in% atac_seqnames,], GenomicRanges::GRanges(seqnames=V1, ranges=IRanges::IRanges(1, V2)))
        gff = as.data.frame(rtracklayer::readGFF(paste0(refdata, "/genes/genes.gtf.gz")))
    } else {
        cs = genome$chromSizes
        cs = cs[GenomeInfoDb::seqnames(cs) %in% atac_seqnames]
        ## cs = GenomicRanges::GRanges(seqnames=as.vector(GenomicRanges::seqnames(genome$chromSizes)),
        ##                             ranges=IRanges::IRanges(0, GenomicRanges::width(genome$chromSizes)))
        ga = ArchR::getGeneAnnotation()
        exons = as.data.frame(ga$exons)
        exons$type = "exon"
        genes = as.data.frame(ga$genes)
        genes$type = "gene"
        gff = rbind(genes, exons)
        gff$gene_name = gff$symbol
    }
    unlink("ArchR", recursive=TRUE)
    tile.list = lapply(list("tile_500.tsv.gz"=500, "tile_5k.tsv.gz"=5000, "tile_50k.tsv.gz"=50000), function(tile_width) {
        tgr = benj::make_tiles(cs, tile_width)
        tgr = unlist(GenomicRanges::subtract(tgr, genome$blacklist, ignore.strand=TRUE))
        GenomicRanges::start(tgr) = GenomicRanges::start(tgr) - 1
        tgr = benj::peak_annotation(tgr, gff=gff)
        ## tgr = as.data.frame(tgr)
        return(tgr)
    })
    for (tname in names(tile.list)) {
        f = gzfile(tname, "w")
        write.table(as.data.frame(tile.list[[tname]])[c("seqnames", "start", "end", "nearestGene", "distToTSS", "peakType")], f, sep="\t", row.names=FALSE, quote=FALSE)
        close(f)
    }
}

do.call(run, get_args())
