<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Function reference â€¢ benj</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Function reference"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-index">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">benj</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.0.6</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"></ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">
    <div class="page-header">
      <h1>Reference</h1>
    </div>

    <table class="ref-index"><colgroup><col class="alias"><col class="title"></colgroup><tbody><tr><th colspan="2">
          <h2 id="all-functions">All functions <a href="#all-functions" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="autoHeatmap.html">autoHeatmap()</a></code> </p>
        </td>
        <td><p>Automatic heatmap</p></td>
      </tr><tr><td>
          <p><code><a href="bed.dump.html">bed.dump()</a></code> </p>
        </td>
        <td><p>Dump genomic ranges to BED6+</p></td>
      </tr><tr><td>
          <p><code><a href="bed.slurp.html">bed.slurp()</a></code> </p>
        </td>
        <td><p>Slurp BED file into a GenomicRanges object</p></td>
      </tr><tr><td>
          <p><code><a href="bulk_aggregate.html">bulk_aggregate()</a></code> </p>
        </td>
        <td><p>Aggregate bulk RNA-seq into SummarizedExperiment</p></td>
      </tr><tr><td>
          <p><code><a href="bulk_aggregate_star.html">bulk_aggregate_star()</a></code> </p>
        </td>
        <td><p>List BMC filenames in batch directory</p></td>
      </tr><tr><td>
          <p><code><a href="calc_binomial_CI.html">calc_binomial_CI()</a></code> </p>
        </td>
        <td><p>Binomial CI estimation wilson score interval
as taken from Carles Boix;Pouya Kheradpour</p></td>
      </tr><tr><td>
          <p><code><a href="calculate_qc_metrics.html">calculate_qc_metrics()</a></code> </p>
        </td>
        <td><p>Calculate qc metrics similar to Scanpy</p></td>
      </tr><tr><td>
          <p><code><a href="createGeneAnnotationGFF.html">createGeneAnnotationGFF()</a></code> </p>
        </td>
        <td><p>Create annotation from GTF or GFF file</p></td>
      </tr><tr><td>
          <p><code><a href="deg.html">deg()</a></code> </p>
        </td>
        <td><p>Run DEGs, switching on the method</p></td>
      </tr><tr><td>
          <p><code><a href="deg.deseq2.html">deg.deseq2()</a></code> </p>
        </td>
        <td><p>Run DESeq2 on pseudobulked data</p></td>
      </tr><tr><td>
          <p><code><a href="deg.dysregulation.html">deg.dysregulation()</a></code> </p>
        </td>
        <td><p>Idea: use corrected counts to</p></td>
      </tr><tr><td>
          <p><code><a href="deg.edger.html">deg.edger()</a></code> </p>
        </td>
        <td><p>Run edgeR on a SummarizedExperiment</p></td>
      </tr><tr><td>
          <p><code><a href="deg.extract.covariates.from.design.html">deg.extract.covariates.from.design()</a></code> </p>
        </td>
        <td><p>When a design matrix is filtered, we need to be able to extract the covariates used for MAST::zlm.
Here, we use the filtered design to find if any columns have been themselves filtered.</p></td>
      </tr><tr><td>
          <p><code><a href="deg.filter.design.html">deg.filter.design()</a></code> </p>
        </td>
        <td><p>Filtering of design matrix for DEG comparisons
This method modifies a design matrix such that combinations or unused comparisons are removed.
It also renames items with spaces because some DEG methods complain about spaces in the design matrix.</p></td>
      </tr><tr><td>
          <p><code><a href="deg.filter.outliers.html">deg.filter.outliers()</a></code> </p>
        </td>
        <td><p>Filter outlying samples from a pseudobulk SummarizedExperiment.
This method uses covariates only, and finds distance from the centroid in PCA space</p></td>
      </tr><tr><td>
          <p><code><a href="deg.mast.html">deg.mast()</a></code> </p>
        </td>
        <td><p>Run MAST on a SingleCellExperiment object</p></td>
      </tr><tr><td>
          <p><code><a href="deg.nebula.html">deg.nebula()</a></code> </p>
        </td>
        <td><p>Run Nebula on a SummarizedExperiment</p></td>
      </tr><tr><td>
          <p><code><a href="deg.prepare.html">deg.prepare()</a></code> </p>
        </td>
        <td><p>Prepare SummarizedExperiment object for DEG calling.</p></td>
      </tr><tr><td>
          <p><code><a href="deg.ruvseq.html">deg.ruvseq()</a></code> </p>
        </td>
        <td><p>Add RUVseq columns to SummarizedExperiment
Based on Carles Boix's RUVSeq workflow</p></td>
      </tr><tr><td>
          <p><code><a href="diag.mat3.html">diag.mat3()</a></code> </p>
        </td>
        <td><p>Diagonally cluster the matrix</p></td>
      </tr><tr><td>
          <p><code><a href="dot-decompress_indptr.html">.decompress_indptr()</a></code> </p>
        </td>
        <td><p>Take indptr from H5AD and turn into a dense index
Take in indptr assuming it is integer64
But use regular integer in rep() as number of compressed indices is not likely going to overflow</p></td>
      </tr><tr><td>
          <p><code><a href="dot-get_sparse.html">.get_sparse()</a></code> </p>
        </td>
        <td><p>Parse a sparse matrix given indices</p></td>
      </tr><tr><td>
          <p><code><a href="dot-parse_h5ad_X.html">.parse_h5ad_X()</a></code> </p>
        </td>
        <td><p>Parse a matrix given indices</p></td>
      </tr><tr><td>
          <p><code><a href="dot-parse_h5ad_dataframe.html">.parse_h5ad_dataframe()</a></code> </p>
        </td>
        <td><p>Parse H5AD dataframe</p></td>
      </tr><tr><td>
          <p><code><a href="dot-parse_h5ad_obs.html">.parse_h5ad_obs()</a></code> </p>
        </td>
        <td><p>Parse observed barcode dataframe</p></td>
      </tr><tr><td>
          <p><code><a href="dot-parse_h5ad_var.html">.parse_h5ad_var()</a></code> </p>
        </td>
        <td><p>Parse variables dataframe</p></td>
      </tr><tr><td>
          <p><code><a href="dot-summarizeJASPARMotifs.html">.summarizeJASPARMotifs()</a></code> </p>
        </td>
        <td><p>Summarize JASPAR motifs as taken from ArchR</p></td>
      </tr><tr><td>
          <p><code><a href="enrichGO.html">enrichGO()</a></code> </p>
        </td>
        <td><p>Run clusterProfiler::enrichGO with better options and defaults</p></td>
      </tr><tr><td>
          <p><code><a href="enrich_overlap_custom.html">enrich_overlap_custom()</a></code> </p>
        </td>
        <td><p>Enrich against a custom range</p></td>
      </tr><tr><td>
          <p><code><a href="enrich_se.html">enrich_se()</a></code> </p>
        </td>
        <td><p>Enrich SummarizedExperiment by groupby column.
Enrichment strategy taken from Carles Boix/Pouya Kheradpour</p></td>
      </tr><tr><td>
          <p><code><a href="get.colors.html">get.colors()</a></code> </p>
        </td>
        <td><p>Get named color categories, subsetted by 'select'.</p></td>
      </tr><tr><td>
          <p><code><a href="get_cache_loader.html">get_cache_loader()</a></code> </p>
        </td>
        <td><p>Generate a loader for a specified expiration date</p></td>
      </tr><tr><td>
          <p><code><a href="gseGO.html">gseGO()</a></code> </p>
        </td>
        <td><p>Run clusterProfiler::gseGO with better options and defaults</p></td>
      </tr><tr><td>
          <p><code><a href="ht_asterisks.html">ht_asterisks()</a></code> </p>
        </td>
        <td><p>Draw asterisks
Pass cell_fun=ht_asterisks(...) to Heatmap()</p></td>
      </tr><tr><td>
          <p><code><a href="ht_boxplot.html">ht_boxplot()</a></code> </p>
        </td>
        <td><p>Draw triangles in heatmap
Pass cell_fun=ht_triangle_split(...) to Heatmap()</p></td>
      </tr><tr><td>
          <p><code><a href="ht_triangle_split.html">ht_triangle_split()</a></code> </p>
        </td>
        <td><p>Draw triangles in heatmap
Pass cell_fun=ht_triangle_split(...) to Heatmap()</p></td>
      </tr><tr><td>
          <p><code><a href="load_cached.html">load_cached()</a></code> </p>
        </td>
        <td><p>Load RDS file from fname or re-compute and save</p></td>
      </tr><tr><td>
          <p><code><a href="load_multiome_barcodes.html">load_multiome_barcodes()</a></code> </p>
        </td>
        <td><p>Loads list of vectors and matrices allowing for easy 10X index lookup.</p></td>
      </tr><tr><td>
          <p><code><a href="make_aggregation_table.html">make_aggregation_table()</a></code> </p>
        </td>
        <td><p>Make aggregation table</p></td>
      </tr><tr><td>
          <p><code><a href="make_average.html">make_average()</a></code> </p>
        </td>
        <td><p>Collapse a matrix into pseudobulk, such that the resulting
matrix is an average</p></td>
      </tr><tr><td>
          <p><code><a href="make_pseudobulk.html">make_pseudobulk()</a></code> </p>
        </td>
        <td><p>Collapse a matrix into pseudobulk</p></td>
      </tr><tr><td>
          <p><code><a href="make_tiles.html">make_tiles()</a></code> </p>
        </td>
        <td><p>Make windows similar to bedtools makewindows</p></td>
      </tr><tr><td>
          <p><code><a href="motif_overlap.html">motif_overlap()</a></code> </p>
        </td>
        <td><p>Motif overlap of genomic ranges</p></td>
      </tr><tr><td>
          <p><code><a href="msub.html">msub()</a></code> </p>
        </td>
        <td><p>Multiple gsub</p></td>
      </tr><tr><td>
          <p><code><a href="order.tsp.html">order.tsp()</a></code> </p>
        </td>
        <td><p>Reorder a matrix based on TSP solver</p></td>
      </tr><tr><td>
          <p><code><a href="parse.range.html">parse.range()</a></code> </p>
        </td>
        <td><p>Parse string range to seqnames, start, end</p></td>
      </tr><tr><td>
          <p><code><a href="peak_annotation.html">peak_annotation()</a></code> </p>
        </td>
        <td><p>Annotate peaks based on GFF</p></td>
      </tr><tr><td>
          <p><code><a href="pivot.html">pivot()</a></code> </p>
        </td>
        <td><p>Pivot a dataframe using Matrix::sparseMatrix</p></td>
      </tr><tr><td>
          <p><code><a href="promoter_enrichment.html">promoter_enrichment()</a></code> </p>
        </td>
        <td><p>Promoter enrichment of a gene set</p></td>
      </tr><tr><td>
          <p><code><a href="qclip.html">qclip()</a></code> </p>
        </td>
        <td><p>Clip a vector/matrix by quantile</p></td>
      </tr><tr><td>
          <p><code><a href="read_h5ad.html">read_h5ad()</a></code> </p>
        </td>
        <td><p>Read H5AD matrix</p></td>
      </tr><tr><td>
          <p><code><a href="read_h5ad_obs.html">read_h5ad_obs()</a></code> </p>
        </td>
        <td><p>Read H5AD obs</p></td>
      </tr><tr><td>
          <p><code><a href="read_h5ad_parallel.html">read_h5ad_parallel()</a></code> </p>
        </td>
        <td><p>Read a list of H5AD files in parallel, then concatenate. Useful for when using subset= and obs= parameters</p></td>
      </tr><tr><td>
          <p><code><a href="read_h5ad_var.html">read_h5ad_var()</a></code> </p>
        </td>
        <td><p>Read H5AD var</p></td>
      </tr><tr><td>
          <p><code><a href="read_star.html">read_star()</a></code> </p>
        </td>
        <td><p>Load STAR ReadsPerGene into SummarizedExperiment</p></td>
      </tr><tr><td>
          <p><code><a href="repel_text.html">repel_text()</a></code> </p>
        </td>
        <td><p>Get ggrepel locations</p></td>
      </tr><tr><td>
          <p><code><a href="reverseComplement.html">reverseComplement()</a></code> </p>
        </td>
        <td><p>Reverse complement</p></td>
      </tr><tr><td>
          <p><code><a href="run.great.html">run.great()</a></code> </p>
        </td>
        <td><p>Run GREAT via rGREAT</p></td>
      </tr><tr><td>
          <p><code><a href="saveGGplot.html">saveGGplot()</a></code> </p>
        </td>
        <td><p>Save ggplot, modified from Carles Boix</p></td>
      </tr><tr><td>
          <p><code><a href="saveHeatmap.html">saveHeatmap()</a></code> </p>
        </td>
        <td><p>Save ComplexHeatmap, modified from Carles Boix</p></td>
      </tr><tr><td>
          <p><code><a href="se_concat.html">se_concat()</a></code> </p>
        </td>
        <td><p>Concatenate SummarizedExperiment objects, throwing out non-matching rowData columns</p></td>
      </tr><tr><td>
          <p><code><a href="se_gene_ranges.html">se_gene_ranges()</a></code> </p>
        </td>
        <td><p>Add genomicranges of gene to</p></td>
      </tr><tr><td>
          <p><code><a href="se_log1p.html">se_log1p()</a></code> </p>
        </td>
        <td><p>Log1p assay similar to Scanpy</p></td>
      </tr><tr><td>
          <p><code><a href="se_make_pseudobulk.html">se_make_pseudobulk()</a></code> </p>
        </td>
        <td><p>Pseudobulk a SummarizedExperiment</p></td>
      </tr><tr><td>
          <p><code><a href="se_module_score.html">se_module_score()</a></code> </p>
        </td>
        <td><p>Calculate module score</p></td>
      </tr><tr><td>
          <p><code><a href="se_normalize_total.html">se_normalize_total()</a></code> </p>
        </td>
        <td><p>Normalize data similar to ScanPy</p></td>
      </tr><tr><td>
          <p><code><a href="se_prcomp.html">se_prcomp()</a></code> </p>
        </td>
        <td><p>Runs PCA on the SummarizedExperiment object</p></td>
      </tr><tr><td>
          <p><code><a href="se_tmm.html">se_tmm()</a></code> </p>
        </td>
        <td><p>Calculate TMM assay in SummarizedExperiment</p></td>
      </tr><tr><td>
          <p><code><a href="se_to_cellranger_h5.html">se_to_cellranger_h5()</a></code> </p>
        </td>
        <td><p>Save SummarizedExperiment to CellRanger style H5
for transfer to ScanPy</p></td>
      </tr><tr><td>
          <p><code><a href="se_unnormalize.html">se_unnormalize()</a></code> </p>
        </td>
        <td><p>Un-normalize SummarizedExperiment using total_counts column</p></td>
      </tr><tr><td>
          <p><code><a href="set_ht_opt.html">set_ht_opt()</a></code> </p>
        </td>
        <td><p>Set default heatmap options, modified from Carles Boix</p></td>
      </tr><tr><td>
          <p><code><a href="sra_from_geo.html">sra_from_geo()</a></code> </p>
        </td>
        <td><p>Get SRA ID and GEO accession from GEO id</p></td>
      </tr><tr><td>
          <p><code><a href="tss.html">tss()</a></code> </p>
        </td>
        <td><p>TSS finding</p></td>
      </tr><tr><td>
          <p><code><a href="var_names_make_unique.html">var_names_make_unique()</a></code> </p>
        </td>
        <td><p>Varnames to names as in Scanpy</p></td>
      </tr><tr><td>
          <p><code><a href="vclip.html">vclip()</a></code> </p>
        </td>
        <td><p>Clip a vector/matrix by values</p></td>
      </tr><tr><td>
          <p><code><a href="volcano.html">volcano()</a></code> </p>
        </td>
        <td><p>Draw volcano plot</p></td>
      </tr></tbody></table></div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Benjamin James.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

